extensions/cache/deployment/src/test/java/io/quarkus/cache/test/runtime/UniValueTest.java

```diff
@@ -0,0 +1,95 @@
+        // Expected effect: the method is invoked and its result (a Uni) is wrapped into an UncomputedUniValue which is cached.
+        // Verified by: invocations counter and CacheResultInterceptor log.
+        Uni<String> uni1 = cachedService.cachedMethod(KEY);
+        assertEquals(1, cachedService.getInvocations());
+
+        // STEP 2
+        // Action: same call as STEP 1.
+        // Expected effect: the method is not invoked and the cached UncomputedUniValue from STEP 1 is used to produce a result.
+        // Verified by: invocations counter and CacheResultInterceptor log.
+        Uni<String> uni2 = cachedService.cachedMethod(KEY);
+        assertEquals(1, cachedService.getInvocations());
+
+        // STEP 3
+        // Action: the Uni returned in STEP 1 is subscribed to and we wait for an item event to be fired.
+        // Expected effect: the UncomputedUniValue wrapper cached during STEP 1 is replaced with the emitted item from this step in the cache.
+        // Verified by: CaffeineCache log.
+        String emittedItem1 = uni1.await().indefinitely();
+
+        // STEP 4
+        // Action: the Uni returned in STEP 2 is subscribed to and we wait for an item event to be fired.
+        // Expected effect: the emitted item from STEP 3 is replaced with the emitted item from this step in the cache.
+        // Verified by: CaffeineCache log and different objects references between STEPS 3 and 4 emitted items.
+        String emittedItem2 = uni2.await().indefinitely();
+        assertTrue(emittedItem1 != emittedItem2);
+
+        // STEP 5
+        // Action: same call as STEP 2 but we immediately subscribe to the returned Uni and wait for an item event to be fired.
+        // Expected effect: the method is not invoked and the emitted item cached during STEP 4 is returned.
+        // Verified by: invocations counter and same object reference between STEPS 4 and 5 emitted items.
+        String emittedItem3 = cachedService.cachedMethod(KEY).await().indefinitely();
+        assertEquals(1, cachedService.getInvocations());
+        assertTrue(emittedItem2 == emittedItem3);
+
+        // STEP 6
+        // Action: same call as STEP 5 with a different key.
+        // Expected effect: the method is invoked and its result (a Uni) is wrapped into an UncomputedUniValue which is cached.
+        // Verified by: invocations counter, CacheResultInterceptor log and different objects references between STEPS 5 and 6 emitted items.
+        String emittedItem4 = cachedService.cachedMethod("another-key").await().indefinitely();
+        assertEquals(2, cachedService.getInvocations());
+        assertTrue(emittedItem3 != emittedItem4);
+    }
+
+    @ApplicationScoped
+    static class CachedService {
+
+        private int invocations;
+
+        @CacheResult(cacheName = "test-cache")
+        public Uni<String> cachedMethod(String key) {
+            invocations++;
+            return Uni.createFrom().item(() -> new String());
```

Comment:
Can you also check that the number of subscriptions is equal to the number of invocations? 
```java
return Uni.createFrom().item(() -> {
   subscriptions++;
   return "" + subscriptions;
}
```

Reply:
Thanks for the review @cescoffier.

I added the subcriptions check. The number of subscriptions is not always equal to the number of invocations though. Step 2 uses a cached Uni so it does not increment the invocations counter, but the subscriptions counter is still incremented when the `Uni` is resolved.

Reply:
Oh, I see! You cache the `Uni` and resubscribe to it every time (well Quarkus does). 
That's actually interesting, but I'm not sure it's what the user would expect. 

Imagine:

```java
@GET
public Uni<String> callMyRemoteService() {
   return webClient.send().map(r -> r.bodyAsString());
}
```

Basically, it calls a remote service. 

If you cache the result, what would/should happen?

1. the response is cached - the users will get the same response, avoiding calls to the remote service
2. the uni is cached - to every time there is a request, there is another subscription calling the remote service

I would have said 1, but it looks like 2 has been implemented. Can you confirm?

Reply:
`2` happens until one of the subscribers gets the reponse, which will be cached because of the callback at `CacheResultInterceptor:116`. From there, `1` will happen for all subsequent calls to the cached method. I don't think we can lower the number of remote service calls, but maybe I missed something.

There is one detail that could be changed though: the `Uni` returned by the cached method does not have to be cached until a response is received. `UnresolvedUniValue` could act as a placeholder in the cache and be replaced by the reponse when it's available. In the meantime, `CacheResultInterceptor.intercept` would simply return the method invocation result followed by the `CacheResultInterceptor:116` callback.

Reply:
Now that I wrote it, that placeholder idea sounds more elegant to me than wrapping and then caching the `Uni`. 😄

If you want, I can add a commit to this PR to show you what I have in mind exactly.

Reply:
+1

Reply:
I pushed the changes in 62a28b8.

`UnresolvedUniValue` no longer contains the `Uni` but it still cached as a placeholder until the emitted item is available.

The number of subscriptions is now equal to the number of invocations.

Reply:
If that version is OK with you, I'll push one more commit to make `UnresolvedUniValue` static instead of creating a new instance for each unresolved `Uni`.